### PR TITLE
Comment liking, recent feed is now recent, cards and comments flagging, extended comments payload.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich ,

Incremental updates from our features list.

This pull request might look cleaner if https://github.com/KnowSheet/CTFO_BE/pull/20 is merged in first.

Thanks!
Dima

### CODE REVIEW

File: server.h [hunk 1]
```diff
@@ -189,158 +509,227 @@ class CTFOServer final {
...
+                }
+              }
+              if (!token_is_valid) {
+                DebugPrint(Printf("[/ctfo/card] Invalid token. Requested URL = '%s'", requested_url.c_str()));
+                return Response("NEED VALID UID-TOKEN PAIR\n", HTTPResponseCode.Unauthorized);
+              } else {
+                DebugPrint(Printf("[/ctfo/card] Token validated. Requested URL = '%s'", requested_url.c_str()));
+                // TODO(dkorolev): Do something smart about non-existing cards.
+                auto cards_mutator = Dictionary<Card>::Mutator(data);
+                cards_mutator.Delete(cid);
```
Comment (mzhurovich) [inline server.h:584]: So everybody can delete any card, not only his own?

Comment (dkorolev) [inline server.h:584]: For now, yes :-) Could you file a bug against me pls?

File: server.h [hunk 2]
```diff
@@ -874,17 +902,31 @@ class CTFOServer final {
...
+    const uint64_t now = static_cast<uint64_t>(bricks::time::Now());
     for (const auto& card : cards) {
-      if (!answers.Has(uid, card.cid)) {
-        candidates.push_back(card.cid);
+      if (!answers.Has(uid, card.cid) && !flagged_cards.Has(card.cid, uid)) {
+        // For the recent feed, relevance is the function of the age of he card.
+        // Added just now => 1.00. Added 24 hour ago => 0.99. Added 48 hours ago => 0.99^2. Etc.
+        const double time_key = std::pow(0.99, (now - card.ms) * (1.0 / (1000 * 60 * 60 * 24)));
+        hot_cards.insert(std::make_pair(RandomDouble(0.2, 0.4), card.cid));
+        recent_cards.insert(std::make_pair(time_key, card.cid));
```
Comment (mzhurovich) [inline server.h]: `.emplace`?

File: server.h [hunk 3]
```diff
@@ -874,17 +902,31 @@ class CTFOServer final {
...
+        if (hot_cards.size() > max_count) {
+          hot_cards.erase(*hot_cards.begin());
```
Comment (mzhurovich) [inline server.h]: `*` looks redundant.

File: server.h (hunk 1 above)
Comment (mzhurovich) [inline server.h:584]: Done.

Comment (dkorolev) [inline server.h:584]: :+1:

File: server.h (hunk 2 above)
Comment (dkorolev) [inline server.h]: :+1: 2597ecd

File: server.h (hunk 3 above)
Comment (dkorolev) [inline server.h]: :+1: 2597ecd

File: server.h [hunk 4]
```diff
@@ -81,101 +82,420 @@ class CTFOServer final {
...
+  void Join() { HTTP(port_).Join(); }
+
+ private:
+  typedef void (CTFOServer::*CTFOServerMemberFunctionServingRequest)(Request);
+
+  std::function<void(Request)> BindToThis(CTFOServerMemberFunctionServingRequest handler) {
+    return std::bind(handler, this, std::placeholders::_1);
+  };
+
+  void RouteAuth(Request r) {
```
Comment (mzhurovich) [inline server.h]: `RouteAuthiOS`?

File: server.h [hunk 5]
```diff
@@ -874,17 +902,31 @@ class CTFOServer final {
...
+    std::set<std::pair<double, CID>> hot_cards;
+    std::set<std::pair<double, CID>> recent_cards;
+    const UID uid = StringToUID(response.user.uid);
+
+    const uint64_t now = static_cast<uint64_t>(bricks::time::Now());
     for (const auto& card : cards) {
-      if (!answers.Has(uid, card.cid)) {
-        candidates.push_back(card.cid);
+      if (!answers.Has(uid, card.cid) && !flagged_cards.Has(card.cid, uid)) {
+        // For the recent feed, relevance is the function of the age of he card.
```
Comment (mzhurovich) [inline server.h]: s/he/the/

File: test.cc [hunk 6]
```diff
@@ -493,6 +554,25 @@ TEST(CTFO, SmokeTest) {
     EXPECT_EQ(1u, my_cards_response.cards[2].number_of_comments);
   }
 
+  // Confirm that the recent feed also mentiones that this card has one comment.
```
Comment (mzhurovich) [inline test.cc]: s/mentiones/mentions/

File: test.cc [hunk 7]
```diff
@@ -837,6 +1060,42 @@ TEST(CTFO, SmokeTest) {
   }
 
   // TODO(dkorolev): Test that I can only delete my own cards.
+
+  // Confirm that the recent feed still contains two of mine cards as the most recently added ones.
```
Comment (mzhurovich) [inline test.cc]: **my** cards?

File: server.h (hunk 4 above)
Comment (dkorolev) [inline server.h]: :+1: fd1e389

File: server.h (hunk 5 above)
Comment (dkorolev) [inline server.h]: :+1: fd1e389

File: test.cc (hunk 6 above)
Comment (dkorolev) [inline test.cc]: :+1: fd1e389

File: test.cc (hunk 7 above)
Comment (dkorolev) [inline test.cc]: :+1: fd1e389

**Comment by mzhurovich** [comment]
@dkorolev LGTM! Merging.

**Pull request merged**